src/asm_state_machine.h

```diff
@@ -71,6 +88,180 @@ static inline void ASM_StateMachine_CompleteTrim(const RedisModuleSlotRangeArray
   slots_tracker_remove_deleted_slots(slots);
 }
 
-#ifdef __cplusplus
+// START KEY SPACE VERSION QUERY TRACKER IMPLEMENTATION
+
+// Define hash map type for tracking query versions -> query counts
+KHASH_MAP_INIT_INT(query_key_space_version_tracker, uint32_t);
+
+// Static hash map instance for tracking query versions
+extern khash_t(query_key_space_version_tracker) *query_key_space_version_map;
+
+// Mutex for thread-safe hash map operations
+extern pthread_mutex_t query_version_tracker_mutex;
+
+static inline void ASM_KeySpaceVersionTracker_Init() {
+  if (query_key_space_version_map != NULL) {
+    kh_destroy(query_key_space_version_tracker, query_key_space_version_map);
+  }
+  query_key_space_version_map = kh_init(query_key_space_version_tracker);
+
+#if ASM_SANITIZER_ENABLED
+  // Initialize sanitizer allocation tracking
+  if (asm_sanitizer_allocs) {
+    free(asm_sanitizer_allocs);
+  }
+  asm_sanitizer_allocs = (int**)malloc(ASM_SANITIZER_INITIAL_CAPACITY * sizeof(int*));
+  asm_sanitizer_alloc_count = 0;
+  asm_sanitizer_alloc_capacity = ASM_SANITIZER_INITIAL_CAPACITY;
+#endif
+  pthread_mutex_init(&query_version_tracker_mutex, NULL);
 }
+
+static inline void ASM_KeySpaceVersionTracker_Destroy() {
+  if (query_key_space_version_map != NULL) {
+    kh_destroy(query_key_space_version_tracker, query_key_space_version_map);
+    query_key_space_version_map = NULL;
+  }
+
+#if ASM_SANITIZER_ENABLED
+  // Clean up any remaining sanitizer allocations
+  if (asm_sanitizer_allocs) {
+    // Do not free the integers, so that sanitizer tests would show leaks
+    free(asm_sanitizer_allocs);
+    asm_sanitizer_allocs = NULL;
+    asm_sanitizer_alloc_count = 0;
+    asm_sanitizer_alloc_capacity = 0;
+  }
+#endif
+  pthread_mutex_destroy(&query_version_tracker_mutex);
+}
+
+static void ASM_KeySpaceVersionTracker_IncreaseQueryCount(uint32_t query_key_space_version) {
+  pthread_mutex_lock(&query_version_tracker_mutex);
+
+  int ret;
+  khiter_t k = kh_put(query_key_space_version_tracker, query_key_space_version_map, query_key_space_version, &ret);
+
+  if (ret == 0) {
+    kh_value(query_key_space_version_map, k)++;
+  } else {
+    kh_value(query_key_space_version_map, k) = 1;
+  }
+
+#if ASM_SANITIZER_ENABLED
+  // Allocate a dummy integer for sanitizer leak detection
+  // We allocate one for each query count increase
+  if (asm_sanitizer_allocs) {
+    // Check if we need to reallocate the array
+    if (asm_sanitizer_alloc_count >= asm_sanitizer_alloc_capacity) {
+      int new_capacity = asm_sanitizer_alloc_capacity * 2;
+      int** new_allocs = (int**)realloc(asm_sanitizer_allocs, new_capacity * sizeof(int*));
```

Comment:
Don't use `malloc` and `free` directly; use `rm_malloc` and `rm_free` family.
We can also assume the `rm_malloc` family [never fails](https://redis.io/docs/latest/develop/reference/modules/modules-api-ref/#redismodule_alloc).